t/integration/test_batches.py

```diff
@@ -270,3 +270,30 @@ def acks(requests: List[SimpleRequest]) -> None:
 
     # After the tasks are done, both results are acked.
     assert acked == [result_1.id, result_2.id]
+
+
+def test_retry(celery_app: Celery, celery_worker: TestWorkController) -> None:
+    """The batch task fails on the first call, is retried and succeeds."""
+
+    def signal(sender: Union[Task, str], **kwargs: Any) -> None:
+        assert celery_app.current_task.name == "t.integration.tasks.retry_if_even"
+
+    # One success, one failure and successful retry
+    counter = SignalCounter(2, signal)
+    signals.task_success.connect(counter)
+
+    result_1 = retry_if_even.delay(1)
+    retry_if_even.delay(2)
+
+    # The flush interval is 1 second and the retry interval is 3 seconds,
+    # this is longer.
+    sleep(5)
+
+    # Let the worker work.
+    _wait_for_ping()
+
+    assert result_1.get() is True
+
+    # The retried task is not accessible here for validation.
```

Comment:
I'm not really sure what you mean -- what would we be validating?

Reply:
I can't do the equivalent of

```
result_2 = retry_if_even.delay(2)
assert result_2.get() is True
```

because that task fails, and an entirely new task is created inside retry_if_even() that is not accessible from this unit test.

Reply:
That makes me a bit concerned that this solution isn't really tenable for retries in the general case. It breaks the contract that the task ID is usable to get results.

I think supporting the countdown / ETA is still a necessarily requirement, but maybe we can figure out how to properly support it in a follow-up?

Reply:
I'm wondering if we can simplify the test significantly. Maybe something like:

1. Call task A
2. Call task B with a countdown of X seconds (longer than the flush interval)
3. Wait for the worker to work.
4. Assert that A processed.
5. Sleep such that flush interval is long enough to process B.
6. Wait for the worker to work.
7. Assert that B was processed.

I think this means we wouldn't even need a new task in the tests?

Reply:
I realized that when re-scheduling a task to retry, the task_id can be specified in apply_async() such that you can still call get() on the original result and the contract is not broken. I've updated the tests and the docs to show this. I would prefer to leave the tests as-is rather than change them as you suggested, since the tests in their current state exercise both the ETA functionality and the retry functionality. Curious to hear your thoughts though.

Reply:
I got the idea for re-using the task_id from looking at how celery itself handles retries internally, which is very similar to what I implemented:

The Task retry() function will create a signature from the original request:
https://github.com/celery/celery/blob/master/celery/app/task.py#L706

Which includes the same task ID:
https://github.com/celery/celery/blob/0a3487b882398ef44d3be34a57b8354e986511af/celery/app/task.py#L607
https://github.com/celery/celery/blob/0a3487b882398ef44d3be34a57b8354e986511af/celery/app/task.py#L115

And then the retry() function will call apply_async() on that signature:
https://github.com/celery/celery/blob/master/celery/app/task.py#L733

As for the mark_as_retry() function, I was mistaken by saying that it's not supported. I think it was working as intended, but its only purpose is to set the state of a task as RETRY in the backend for tracking purposes. If mark_as_retry() is not used, the task stays in the PENDING state, which also works. The other missing piece of the puzzle was the ETA and task_id support.

I thought all this made sense, and I can add test cases for pure ETA requests in addition to the retry test cases.

Would this approach still be acceptable?

Reply:
Thanks for the cross-reference. I'll take another look.

Reply:
> I thought all this made sense, and I can add test cases for pure ETA requests in addition to the retry test cases.

I think the retry test case doesn't really belong here though -- it isn't testing celery-batches, it is testing application code. I think it is clearer to just call `apply_async()` with a countdown and with an `eta` in tests directly.

Looking at the `Task.retry()` function I wonder if we should wrap some of that logic up into a nice method though. I think that could make a nice follow-up to give the incoming requests an easy `retry()` method?

Reply:
Thanks for the second review. That's a good point, the unit test relies more on application behaviour, which we can't guarantee/control in the outside world.

Reply:
I made this change in 69522d25b03d726fdcb912c361b84e713f55c3b2, by the way. I think it is a bit clearer.